### DISCUSSION TITLE
Header files not recognized in project after adding custom library to VCPkg

### DISCUSSION THREAD

**Discussion by drathin4** [microsoft/vcpkg#27775] [Q&A]
I have added a custom library `fleck` to `vcpkg` using the `manifests/registries` feature on `Windows` platform. I have been in in touch with the support team regarding this and have initiated discussions: #27636, #27573, #27308 , #27528.

I used the following tutorial for the same:
[https://devblogs.microsoft.com/cppblog/registries-bring-your-own-libraries-to-vcpkg/](url)

But after the integration of this library, I have 3 issues in all:

1. Problem identifying header (`.h`) files in the project
I have created `vcpkg-registry` at Azure location. I created a project which has `vcpkg.json` & `vcpkg-configuration.json`, `CMakeLists.txt`, `main.cpp` files in the root directory. Later, I used `cmake` build commands to configure/generate/build the project. The project was successfully built and `vcpkg_installed` folder appeared in the `build` directory of the project. It has all the files and folders under `vcpkg_installed\x64-windows`. But in my `main.cpp`, I cannot `#include` the header files. 

![image](https://user-images.githubusercontent.com/79855313/201360884-1fac93b9-5517-4b26-9a69-766b56128089.png)

The `fleck` library that I added has all the header files present under `build\vcpkg_installed\x64-windows\include\fleck`. I have created 3 `static` libraries (`targets`) - `cove.lib`, `nimbus.lib`, `sisyphus.lib` from `fleck` and they're present in `fleck\cove`, `fleck\nimbus`, `fleck\sisyphus` directories. Why are the header files not recognizable ?

I have already done `vcpkg integrate install` when I installed `vcpkg`. Also, I'm able to `#include *.h` files from other dependencies that got added to my project inadvertantly (mentioned in `point 3`), but not from `fleck`.

2. Custom library is not present in `ports`, `versions`, `installed` directories, but only in `buildtrees`, `downloads`, `packages` directories (`VCPKG_ROOT`)
I have created `vcpkg-registry` at Azure location. I created a local project which has `vcpkg.json` & `vcpkg-configuration.json`, `CMakeLists.txt`, `main.cpp` files in the root directory. I used the command `vcpkg install --feature-flags=registries,manifests` from the project location and the library was added _partially_ to `vcpkg`. It also created a `vcpkg_installed` folder in the project's root directory. The library was not added to the folders/directories of `VCPKG_ROOT` mentioned above. What is the reason for that ? Is it the case that only open source libraries available in VCPkg search are added at all locations ?

3. Unnecessary dependencies getting added to my project
For my `fleck` library, the only dependency that I have added is `vectorclass` in `vcpkg.json`. However, when I use either `vcpkg install --feature-flags=registries,manifests` or `cmake -S . -B .\build`, all the existing libraries from `VCPKG_ROOT` get installed in the sub-folders of `vcpkg_installed\x64-windows` directory. They're shown in the attached `log.txt` file. What could be the reason for that ?

For all these queries, I have attached the following files: `portfile.cmake`, `vcpkg.json` from `vcpkg-registry`, `CMakeLists.txt` from Azure cloud location, `vcpkg.json`, `vcpkg-configuration.json`, `CMakeLists.txt` from project root directory, `log.txt` illustrating the library installation and project building log. Hoping for your reply. Thanks.

[CMakeLists_azure.txt](https://github.com/microsoft/vcpkg/files/10020573/CMakeLists_azure.txt)
[portfile.cmake.txt](https://github.com/microsoft/vcpkg/files/9990977/portfile.cmake.txt)
[vcpkg_registry.json.txt](https://github.com/microsoft/vcpkg/files/9990978/vcpkg_registry.json.txt)
[log.txt](https://github.com/microsoft/vcpkg/files/9990994/log.txt)
[vcpkg_project.json.txt](https://github.com/microsoft/vcpkg/files/9990986/vcpkg_project.json.txt)
[vcpkg-configuration.json.txt](https://github.com/microsoft/vcpkg/files/9990988/vcpkg-configuration.json.txt)
[CMakeLists_project.txt](https://github.com/microsoft/vcpkg/files/10020580/CMakeLists_project.txt)

**Comment by BillyONeal** [comment]
>But in my main.cpp, I cannot #include the header files.

You show a screenshot but not build output. Is the problem that *intellisense* is not finding the thing or that your actual downstream consuming build is not?

>Custom library is not present in ports, versions, installed directories, but only in buildtrees, downloads, packages directories (VCPKG_ROOT)

That is expected, because you got the thing from a different registry.

>Unnecessary dependencies getting added to my project
>For my fleck library, the only dependency that I have added is vectorclass in vcpkg.json.

Your repro says:

```
  "dependencies": [
	"vectorclass",
    "fleck"
  ],
```

What are the dependencies of `fleck`?

**Reply by drathin4** [reply]
> You show a screenshot but not build output. Is the problem that intellisense is not finding the thing or that your actual downstream consuming build is not?

Please check the build output in the attached `log.txt` file. I think, the problem is not just with the intellisense, if I `#include *.h` files forcefully, I get build errors.

> What are the dependencies of fleck?

Please check the fleck dependencies in the attached `vcpkg_registry.json.txt` file. There's just 1 dependency - `vectorclass`.

Note that I have used 2 `vcpkg.json` files - one in `vcpkg-registry` (attached as `vcpkg_registry.json.txt` in this thread) and other in `project root directory` (attached as `vcpkg_project.json.txt` in this thread). The one that you showed is from the project root directory. Also, there are 2 `CMakeLists.txt` files - one at `portfile.cmake URL` where the library is present (attached as `CMakeLists_azure.txt` in this thread) and other in `project root directory` (attached as `CMakeLists_project.txt` in this thread).

I have used the command `cmake -S . -B ./build` to build the project because I have used `set(CMAKE_TOOLCHAIN_FILE $ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake)` in `CMakeLists.txt` of `project root directory`.

For integrating this library, I used the following tutorial:
[https://devblogs.microsoft.com/cppblog/registries-bring-your-own-libraries-to-vcpkg/](url).

**Reply by JavierMatosD** [reply]
Hi @drathin4, I'm not seeing anything in the files you included that would explain your issue. I am going to try and get a second pair of eyes. In the meantime, which headers are you unable to include, and what build errors do you get?

**Comment by drathin4** [comment]
Hello @JavierMatosD, I was able to integrate the library in `vcpkg` . I made some changes in `CMakeLists.txt` file and now I'm able to include the files. However, I still have a few questions:

1. Although I can include the files, I have to mention the full path in it. For eg. `#include "build/vcpkg_installed/x64-windows/include/fleck/cove/silo/cv_file.h"`. Is this expected ?
2. This is the same `point 3` mentioned in the above discussion (dated `11th Nov`) regarding unnecessary dependencies getting added to the project. Any clue to that ?
3. This is a generic problem that I've had since past few months regarding `pybind11` library, not specific to this integration. If required, I can create a separate `vcpkg discussion` on this. When I add `pybind11` using `vcpkg`, I can't build `VisualStudio` in `debug configuration`. The error it gives out is: **LINK : fatal error LNK1104: cannot open file 'python310.lib'**. I checked in the project properties (`Linker->Input->AdditionalDependencies`) for `debug configuration`, and the binary file is correctly populated as `python310_d.lib`, but despite of that, `VisualStudio` queries the `python310.lib`. Is this a bug in `vcpkg` ? When I create a `cmake` project having `pybind11` dependency without using `vcpkg`, the `debug configuration` builds correctly in `VisualStudio`.

**Reply by JavierMatosD** [reply]
Hi Rathin, 

> Although I can include the files, I have to mention the full path in it. For eg. #include "build/vcpkg_installed/x64-windows/include/fleck/cove/silo/cv_file.h". Is this expected ?

You should not have to include the full path. You should be able to just #include "fleck/cove/silo/cv_file.h" or something similar. Make sure that the list file at the registry level includes something like the following:

```target_include_directories(fleck::fleck INTERFACE $<INSTALL_INTERFACE:include>)```

If this is still an issue, please send me the updated listfile for the registry. 

> This is the same point 3 mentioned in the above discussion (dated 11th Nov) regarding unnecessary dependencies getting added to the project. Any clue to that ?

Your repro states: 
```
  "dependencies": [
	"vectorclass",
    "fleck"
  ],
```
You've stated that fleck depends on vectorclass. This means that just declaring "fleck" as a dependency should suffice. As for the extra dependencies, go to C:\Users\SEZ5505\AppData\Local\vcpkg\registries\git-trees\85e5a9a4ff38e5312e2b4bdc692871de836579cd location and inspect the vcpkg.json to make sure it's not an older vcpkg.json from a previous edit. If you look at the log.txt file, you will see something like 

``` 
  * boost-algorithm[core]:x64-windows -> 1.80.0
  * boost-align[core]:x64-windows -> 1.80.0
  * boost-any[core]:x64-windows -> 1.80.0
  * boost-array[core]:x64-windows -> 1.80.0
  .
  . 
  .
 ```
The '*' before the port name indicates that these are transitive dependencies. I recall from a previous discussion that you were trying to include ```qt``` and ```boost``` stuff. It could be that C:\Users\SEZ5505\AppData\Local\vcpkg\registries\git-trees\85e5a9a4ff38e5312e2b4bdc692871de836579cd is the wrong reference.

> This is a generic problem that I've had since past few months regarding pybind11 library, not specific to this integration. If required, I can create a separate vcpkg discussion on this. When I add pybind11 using vcpkg, I can't build VisualStudio in debug configuration. The error it gives out is: LINK : fatal error LNK1104: cannot open file 'python310.lib'. I checked in the project properties (Linker->Input->AdditionalDependencies) for debug configuration, and the binary file is correctly populated as python310_d.lib, but despite of that, VisualStudio queries the python310.lib. Is this a bug in vcpkg ? When I create a cmake project having pybind11 dependency without using vcpkg, the debug configuration builds correctly in VisualStudio.

I am not sure about this. I need to do some digging. The pybind11 repository has recently been updated with some bug fixes. Some of them are related to vcpkg. It is possible that simply updating the port could fix the problem. That said, this is reproducible. Please send me repro steps :)

Additional notes:
The following in your CMakeLists.txt (project) is no longer required:

```
set(VCPKG_FEATURE_FLAGS "manifests")
list(APPEND VCPKG_FEATURE_FLAGS "registries")
```
This won't cause you any issues, but it just doesn't do anything.

**Reply by drathin4** [reply]
Thanks @JavierMatosD  for your response. I was able to resolve `queries 1, 2`. I'll give you the files/steps necessary for reproducing `query 3`. 

Files: 
[CMakeLists.txt](https://github.com/microsoft/vcpkg/files/10228116/CMakeLists.txt)
[vcpkg.json.txt](https://github.com/microsoft/vcpkg/files/10228225/vcpkg.json.txt)
[vcpkg-configuration.json.txt](https://github.com/microsoft/vcpkg/files/10228226/vcpkg-configuration.json.txt)

Build steps:
cmake -S . -B .\build -D CMAKE_BUILD_TYPE=Release
cmake --build .\build --config Release
cmake --build .\build --config Debug
If I build in `debug` configuration, the error is: **LINK : fatal error LNK1104: cannot open file 'python310.lib'**
Just as a reminder, I'm using `VisualStudio` generator on `Windows` platform.

Having said this, I'm facing a new problem now. In my Python binding, I'm using `PYBIND11_MODULE` to call C++ code from Python script. In the script, I'm importing Python's `tkinter module`. When I do that, the Python command line application gives the error: 
**import _tkinter # If this fails your Python may not be configured for Tk
ModuleNotFoundError: No module named '_tkinter'**

I searched for this issue on Google and found a relevant link:
[https://github.com/microsoft/vcpkg/issues/25427](url)
I also checked in my local `portfile.cmake` file of `python3` port and found `"/p:IncludeTkinter=false"`
Can I explicitly enable it in that file or is there any workaround to this ?

PS: I am able to get the python binding of fleck library working without using vcpkg. But I want to integrate `fleck` in `vcpkg` and use additional dependencies of `pybind11, python3.10, vectorclass` through `vcpkg` itself.

Thanks in advance !

**Reply by JavierMatosD** [reply]
I believe the immediate solution for your scripted builds would be to use separate build directories for Debug and Release. The find python doesn't work well with multi config generators. However, for your personal IDE builds, I recommend using VS's open folder instead of using the VS generator. VS will know what to do. You can follow these instructions https://learn.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-170 Finally, if this doesn't work for you then I would appreciate it if you emailed me a ``` --trace ``` run of your configure step.

**Reply by Osyotr** [reply]
You should define `Py_DEBUG` to force pybind11 to use debug version of Python.

**Reply by drathin4** [reply]
@Osyotr: Does this have to be done at the `cmake` level before configuring the build ? I wrote `#define Py_DEBUG` in the `main.cpp` file and that didn't work. I have also created a different build directory `build_d` just for `debug` configuration as suggested by @JavierMatosD, but that hasn't worked either. 

I found a relevant link: [https://github.com/pybind/pybind11/pull/1438](url) in which they have written **-DPy_DEBUG**, so it seems that it's supposed to be a `cmake` command.

**Reply by Osyotr** [reply]
It's a preprocessor definition, you must either set it in project properties or make sure it's defined before any of pybind11 headers.

**Reply by drathin4** [reply]
I wrote:

>#if _DEBUG
#define Py_DEBUG
#endif

as the first 3 lines in `main.cpp` file before including `pybind11` headers as you suggested.

That works for both `Release` & `Debug`. Thanks a lot for your help!
